packages/blade/src/components/Typography/Heading/Heading.tsx

```diff
@@ -0,0 +1,85 @@
+import type { ReactElement } from 'react';
+import type { TextTypes } from '../../../tokens/theme/theme';
+import getPlatformType from '../../../utils/getPlatformType';
+import type { Theme } from '../../BladeProvider';
+import BaseText from '../BaseText';
+import type { BaseTextProps } from '../BaseText/BaseText';
+
+type HeadingCommonProps = {
+  type: TextTypes;
+  children?: React.ReactNode;
+};
+
+type HeadingNormalVariant = HeadingCommonProps & {
+  variant: 'small' | 'medium' | 'large';
+  weight: keyof Theme['typography']['fonts']['weight'];
+};
+
+type HeadingSubHeadingVariant = HeadingCommonProps & {
+  variant: 'subheading';
+  weight: keyof Pick<Theme['typography']['fonts']['weight'], 'bold'>;
+};
+
+/**
+ * Conditionally changing props based on variant.
+ * Overloads or union gives wrong intellisense.
+ */
+export type HeadingProps<T> = T extends {
+  variant: infer Variant;
+}
+  ? Variant extends 'small' | 'medium' | 'large'
+    ? HeadingNormalVariant
+    : Variant extends 'subheading'
+    ? HeadingSubHeadingVariant
+    : T
+  : T;
+
+const getProps = <T extends { variant: 'small' | 'medium' | 'large' | 'subheading' }>({
```

Comment:
Question: 

Why are we extracting out this `getProps` function instead of inlining it directly on the component? Feels like if we are not using it else where better to keep it inline.

Reply:
just kept it as a pattern for readability, at few places we do some modifications inside this function. Since this function is responsible as an abstraction for our shared styles and conditional styles, felt it would be better to keep it out from the Component function